plugins/plugin-chart-scatter/src/Scatter/transformProps.ts

```diff
@@ -140,24 +147,60 @@ export default function transformProps(
 
   const colorFn = CategoricalColorNamespace.getScale(colorScheme as string);
 
-  const sourceDataSet: OptionSourceDataArrayRows = rawData.map(
-    (datum: DataRecord) =>
-      [
-        datum[xField],
-        datum[yField],
-        datum[sizeField] || DEFAULT_BUBBLE_SIZE,
-        ...groupby.map(group => getSeriesName(datum[group as string])),
-      ] as OptionDataValue[],
-  );
+  const sourceDataSet: OptionSourceDataArrayRows = rawData.map((datum: DataRecord) => {
+    const clustering =
+      clusteringEntity != null && !isAggMode
+        ? [getSeriesName(datum[clusteringEntity])]
+        : groupby.map(group => getSeriesName(datum[group as string]));
+    return [
+      datum[xField],
+      datum[yField],
+      datum[sizeField] || DEFAULT_BUBBLE_SIZE,
+      ...clustering,
+    ] as OptionDataValue[];
+  });
+
+  const sourceDataSetDimension = sourceDataSet[0]?.length ? sourceDataSet[0].length : 0;
+
+  const allGroups = rawData.map(datum => {
+    if (clusterType === 'Cluster by Entity' && enableClustering && clusterEntity && !isAggMode) {
+      return getSeriesName(datum[clusterEntity as string]);
+    }
+    return getSeriesName(datum[groupby[0]]);
+  });
 
-  const allGroups = rawData.map(datum => getSeriesName(datum[groupby[0]]));
   const uniqueGroups = Array.from(new Set(allGroups).values());
+  let visualMap;
+
+  if (clusterType === 'hierarchical kMeans' && enableClustering && !isAggMode) {
```

Comment:
Ich würde vorschlagen, die ganze Logik für die VisualMap in eine Funktion auszulagern